src/sage/crypto/key_exchange/diffie_hellman.py

```diff
@@ -0,0 +1,338 @@
+r"""
+Diffie-Hellman Public Key Exchange Scheme
+
+This module contains a toy implementation of the Diffie-Hellman public key
+exchange scheme.
+
+AUTHORS:
+
+- Vincent Macri (2024-07-17): initial version
+"""
+# ****************************************************************************
+#       Copyright (C) 2024 Vincent Macri <vincent.macri@ucalgary.ca>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#                  https://www.gnu.org/licenses/
+# ****************************************************************************
+
+from sage.misc.superseded import experimental
+
+from sage.crypto.key_exchange.key_exchange import KeyExchangeScheme
+
+from sage.arith.misc import is_prime
+from sage.misc.prandom import randint
+from sage.rings.integer import Integer
+from sage.rings.finite_rings.finite_field_constructor import GF
+from sage.rings.finite_rings.finite_field_prime_modn import \
+    FiniteField_prime_modn
+from sage.rings.finite_rings.integer_mod import IntegerMod_abstract
+from sage.structure.proof.proof import WithProof
+
+from typing import Union
+
+class DiffieHellman(KeyExchangeScheme):
+
+    @experimental(37305)
+    def __init__(self, p: Integer, g: Union[Integer, IntegerMod_abstract],
```

Comment:
The conditions on `g` are actually looser than the type annotation implies.

For instance, `DiffieHellman(7, 1/2)` is allowed because 2 has an inverse in GF(7). I'm not sure what the best way to deal with this is. I expect most users using this class will either pass in an integer for g or *maybe* an element of GF(p), which is why I only included these in the annotation. I'm not sure if there is an easy way to do an annotation for "anything that is coercable to an element of GF(p)". I mention in the docstring that `g` needs to be coercable to an element of `GF(p)`.

If people have thoughts on this I'm open to changing this.

Reply:
I don't think this problem inherits from how you constructed this piece of code and since this is a toy framework, I don't think this would be a problem.

If you really want to check this maybe you can try to play with

```py
if isinstance(g, Integer) or isinstance(g, IntegerMod_abstract):
    blah
```
but to me, this is ok.

Reply:
I don't think I want to enforce the typing annotations, I'm happy with `DiffieHellman(7, 1/2)` being allowed. I guess one way to look at it is the type annotations tell you the types that the API *guarantees* to work, i.e. we could decide not to allow rationals later without a deprecation cycle (not that the deprecation rules apply now anyway since it's marked experimental).

Especially since this class is primarily for pedagogical purposes, I lean a lot more to "allow anything (or at least don't enforce any restrictions) that won't throw an error".

That's why I didn't do anything to prevent something like `DiffieHellman(7, GF(9)(14))`. That will work (and respects the type annotations). 14 will be reduced mod 3 to 2, and the `__init__` in DiffieHellman will convert 2 mod 3 to 2 mod 7. I don't know *why* a user would do this, but since the class is for primarily pedagogical use I don't want to stop a user from trying anyway to answer the question "what if I do this?".

Reply:
Sounds *rational* to me (please laugh). I lean more towards not enforcing everything too, as long as it's reasonable it's okay.